ui/StatusQ/src/keychain.mm

```diff
@@ -0,0 +1,262 @@
+        qWarning() << "another local authentication request in progress";
+        return Keychain::StatusGenericError;
+    }
+
+    *context = [[LAContext alloc] init];
+    NSError *authError = nil;
+
+    // Check if Biometrics Authentication is available
+    if (![*context canEvaluatePolicy:authPolicy error:&authError]) {
+        qWarning() << "biometric authentication not available:"
+                   << QString::fromNSString(authError.localizedDescription);
+        return convertError(authError);
+    }
+
+    QEventLoop loop;
+    auto loopPtr = &loop;
+    __block NSError *callbackError = nil;
+    __block BOOL success = NO;
+
+    // Prompt for biometrics authentication
+    [*context evaluatePolicy:authPolicy
+             localizedReason:reason.toNSString()
+                       reply:^(BOOL authSuccess, NSError *error) {
+                           success = authSuccess;
+                           callbackError = error ? [error copy] : nil;
+                           loopPtr->quit();
+                       }];
+
+    // Wait for biometrics authentication finished
+    loop.exec();
+
+    if (!success && callbackError) {
+        qWarning() << "authentication failed:"
+                   << QString::fromNSString(callbackError.localizedDescription);
+        return convertError(callbackError);
+    }
+
+    return Keychain::StatusSuccess;
+}
+
+void Keychain::requestSaveCredential(const QString &account, const QString &password)
+{
+    if (m_future.isRunning()) {
+        return;
+    }
+
+    m_future = QtConcurrent::run([this, account, password]() {
+        setLoading(true);
+        const auto status = saveCredential(account, password);
+        emit saveCredentialRequestCompleted(status);
+        setLoading(false);
+    });
+}
+
+void Keychain::requestDeleteCredential(const QString &account)
+{
+    if (m_future.isRunning()) {
+        return;
+    }
+
+    m_future = QtConcurrent::run([this, account]() {
+        setLoading(true);
+        const auto status = deleteCredential(account);
+        emit deleteCredentialRequestCompleted(status);
+        setLoading(false);
+    });
+}
+
+void Keychain::requestGetCredential(const QString &account)
+{
+    if (m_future.isRunning()) {
+        return;
+    }
+
+    m_future = QtConcurrent::run([this, account]() {
+        setLoading(true);
+        QString credential;
+        const auto status = getCredential(account, &credential);
+        emit getCredentialRequestCompleted(status, credential);
+        setLoading(false);
+    });
+}
+
+void Keychain::cancelActiveRequest()
+{
+    if (m_activeAuthContext != nullptr)
+        [m_activeAuthContext invalidate];
+}
+
+// ContextCleaner takes a pointer to LAContext* and nullifies it on destruction
+class ContextCleaner
+{
+public:
+    explicit ContextCleaner(LAContext **context)
+        : m_context(context)
+    {}
+    ~ContextCleaner() { *m_context = nullptr; };
+
+private:
+    LAContext **m_context;
+};
```

Comment:
There is no need to define own type for that purpose because Qt's `QScopedValueRollback` class can be used.

Reply:
Noice, I thought that Qt has something, but couldn't find it 👍